File: shared/src/androidTest/kotlin/com/commuto/interfacemobile/kmService/KMServiceTest.kt
package com.commuto.interfacemobile.kmService

import androidx.test.core.app.ApplicationProvider
import com.commuto.interfacemobile.db.DatabaseDriverFactory
import com.commuto.interfacemobile.dbService.DBService
import com.commuto.interfacemobile.kmService.kmTypes.KeyPair
import com.commuto.interfacemobile.kmService.kmTypes.PublicKey
import java.util.*
import kotlin.test.Test
import kotlin.test.BeforeTest

internal class KMServiceTest {

    private var driver = DatabaseDriverFactory(ApplicationProvider.getApplicationContext())
    private var dbService = DBService(driver)
    private val kmService = KMService(dbService)

    @BeforeTest
    fun testCreateTables() {
        dbService.createTables()
    }

    @Test
    fun testGenerateAndRetrieveKeyPair() {
        val keyPair: KeyPair = kmService.generateKeyPair()
        val retrievedKeyPair: KeyPair? = kmService.getKeyPair(keyPair.interfaceId)
        assert(Arrays.equals(keyPair.interfaceId, retrievedKeyPair!!.interfaceId))
        assert(keyPair.keyPair.public.equals(retrievedKeyPair.keyPair.public))
        assert(keyPair.keyPair.private.equals(retrievedKeyPair.keyPair.private))
    }

    @Test
    fun testStoreAndRetrievePubKey() {
        val keyPair: KeyPair = kmService.generateKeyPair(storeResult = false)
        val publicKey = PublicKey(keyPair.keyPair.public)
        kmService.storePublicKey(publicKey)
        val retrievedPublicKey: PublicKey? = kmService.getPublicKey(publicKey.interfaceId)
        assert(Arrays.equals(publicKey.interfaceId, retrievedPublicKey!!.interfaceId))
        assert(publicKey.publicKey.equals(retrievedPublicKey.publicKey))
    }

}

File: iosApp/iosAppTests/KMServiceTest.swift
//
//  KMServiceTest.swift
//  iosAppTests
//
//  Created by jimmyt on 11/8/21.
//  Copyright © 2021 orgName. All rights reserved.
//

import Foundation
import XCTest
@testable import iosApp

class KMServiceTest: XCTestCase {
    
    let dbService: DatabaseService = DatabaseService()
    var kmService: KMService?
     

    override func setUpWithError() throws {
        // Put setup code here. This method is called before the invocation of each test method in the class.
        kmService = KMService(dbService: dbService)
        try dbService.connectToDb()
        try dbService.createTables()
    }

    override func tearDownWithError() throws {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
    }
    
    func testGenerateAndRetrieveKeyPair() throws {
        let keyPair = try kmService!.generateKeyPair()
        let retrievedKeyPair = try kmService!.getKeyPair(interfaceId: keyPair.interfaceId)
        XCTAssertEqual(keyPair.interfaceId, retrievedKeyPair!.interfaceId)
        XCTAssertEqual(keyPair.publicKey, retrievedKeyPair!.publicKey)
        XCTAssertEqual(keyPair.privateKey, retrievedKeyPair!.privateKey)
    }
    
    func testStoreAndRetrievePublicKey() throws {
        let keyPair = try kmService!.generateKeyPair()
        let publicKey = try PublicKey(publicKey: keyPair.publicKey)
        try kmService!.storePublicKey(pubKey: publicKey)
        let retrievedPublicKey = try kmService!.getPublicKey(interfaceId: publicKey.interfaceId)
        XCTAssertEqual(publicKey.interfaceId, retrievedPublicKey?.interfaceId)
        XCTAssertEqual(publicKey.publicKey, publicKey.publicKey)
    }

}
